File: 2_Inventory_Model/1_Model/Optimizer/src/update_pi_markov.m
%update_pi: Calculation of PI_markov

function PI_markov = update_pi_markov(markovmodel,prices,t,PI,k,solution_method)

NoP=numel(prices); %Number of Prices
PI_markov=cell(1,2);
R1=markovmodel{1}; %probability transition matrix for R1
R2=markovmodel{2}; %probability transition matrix for R2
if t==1 %for t=1 the belief update values are given
    PI_markov{1}=PI(1);
    PI_markov{2}=PI(2);
    return
end

PI_markov1=zeros(NoP^(t-1), NoP);
PI_markov2=zeros(NoP^(t-1), NoP);
PI_past_R1=PI{t-1}{1}; %belief updates in t-1 for R1
PI_past_R2=PI{t-1}{2}; %belief updates in t-1 for R2
if t==2 % In t=1, pi=[0.5 0.5]. To use only one loop for all t>=2, we have to blow up the size of the belief in period 1 by copying its value 5 (NoP) times
    PI_past_R1=repmat(PI_past_R1,1,NoP);
    PI_past_R2=repmat(PI_past_R2,1,NoP);
end
i=1;
j=1;
while j<=NoP^(t-1)
    for last_price=1:NoP %p_past is the price in the last period p_(t-1)
       for p_t=1:NoP %p_t is the price in the actual period t
           if strcmp(solution_method, 'CEC')  
                    PI_markov1(j,p_t)=PI{1}{1};
                    PI_markov2(j, p_t)=PI{1}{2};
           elseif strcmp(solution_method, 'R1')
                     PI_markov1(j,p_t)=1;
                     PI_markov2(j,p_t)=0;
           elseif strcmp(solution_method, 'R2')
                     PI_markov1(j,p_t)=0;
                     PI_markov2(j,p_t)=1;
           else 
          PI_markov1(j,p_t) = (PI_past_R1(i,last_price)*k(1,1)*R1(last_price, p_t)...
              +PI_past_R2(i,last_price)*k(1,2)*R2(last_price,p_t))... %enumerator for R1
              /(PI_past_R1(i,last_price)*R1(last_price, p_t)+PI_past_R2(i,last_price)*R2(last_price, p_t)); %denominator for R1
          PI_markov2(j, p_t) = (PI_past_R1(i,last_price)*k(2,1)*R1(last_price, p_t)... 
              +PI_past_R2(i,last_price)*k(2,2)*R2(last_price, p_t))...%enumerator for R2
              /(PI_past_R1(i,last_price)*R1(last_price, p_t)+(PI_past_R2(i,last_price)*R2(last_price, p_t))); %denominator for R1
          if isnan(PI_markov1(j, p_t)) %Division by zero -> NO belief update
              PI_markov1(j, p_t)=PI_past_R1(i,last_price);
          end
          if isnan(PI_markov2(j, p_t)) %Division by zero -> NO belief update
              PI_markov2(j, p_t)=PI_past_R2(i,last_price);
          end
           end %neu
       end
       j=j+1;
    end
    i=i+1;
end

PI_markov={PI_markov1, PI_markov2};

File: 2_Inventory_Model/1_Model/Optimizer/src_py/update_pi_markov.py
import numpy as np
from numba import njit

CEC = 1
R1 = 2
R2 = 3


@njit(cache=True)
def _update_pi_markov_core(R1_mat, R2_mat, NoP, t, PI_past_R1, PI_past_R2,
                           k, method_code, initial_pi_r1, initial_pi_r2):
    """Core Bayesian update loop for Markov model (t >= 2).

    Parameters
    ----------
    R1_mat, R2_mat : np.ndarray (NoP, NoP)
        Transition matrices.
    NoP : int
        Number of prices.
    t : int
        Current period (1-based).
    PI_past_R1, PI_past_R2 : np.ndarray
        Previous period belief matrices.
    k : np.ndarray (2, 2)
        Regime transition probabilities.
    method_code : int
        Solution method code.
    initial_pi_r1, initial_pi_r2 : float
        Initial beliefs for CEC fallback.

    Returns
    -------
    PI_markov1, PI_markov2 : np.ndarray, shape (NoP^(t-1), NoP)
    """
    result_rows = 1
    for _ in range(t - 1):
        result_rows *= NoP

    PI_markov1 = np.zeros((result_rows, NoP))
    PI_markov2 = np.zeros((result_rows, NoP))

    i = 0  # index for PI_past (0-based)
    j = 0  # index for result (0-based)

    while j < result_rows:
        for last_price in range(NoP):
            for p_t in range(NoP):
                if method_code == CEC:
                    PI_markov1[j, p_t] = initial_pi_r1
                    PI_markov2[j, p_t] = initial_pi_r2
                elif method_code == R1:
                    PI_markov1[j, p_t] = 1.0
                    PI_markov2[j, p_t] = 0.0
                elif method_code == R2:
                    PI_markov1[j, p_t] = 0.0
                    PI_markov2[j, p_t] = 1.0
                else:
                    # Bayesian update
                    denom = (PI_past_R1[i, last_price] * R1_mat[last_price, p_t]
                             + PI_past_R2[i, last_price] * R2_mat[last_price, p_t])

                    if denom == 0.0:
                        PI_markov1[j, p_t] = PI_past_R1[i, last_price]
                        PI_markov2[j, p_t] = PI_past_R2[i, last_price]
                    else:
                        num_r1 = (PI_past_R1[i, last_price] * k[0, 0] * R1_mat[last_price, p_t]
                                  + PI_past_R2[i, last_price] * k[0, 1] * R2_mat[last_price, p_t])
                        num_r2 = (PI_past_R1[i, last_price] * k[1, 0] * R1_mat[last_price, p_t]
                                  + PI_past_R2[i, last_price] * k[1, 1] * R2_mat[last_price, p_t])
                        PI_markov1[j, p_t] = num_r1 / denom
                        PI_markov2[j, p_t] = num_r2 / denom

                        # NaN check (already handled by denom==0 above, but keep for safety)
                        if np.isnan(PI_markov1[j, p_t]):
                            PI_markov1[j, p_t] = PI_past_R1[i, last_price]
                        if np.isnan(PI_markov2[j, p_t]):
                            PI_markov2[j, p_t] = PI_past_R2[i, last_price]
            j += 1
        i += 1

    return PI_markov1, PI_markov2


def update_pi_markov(R1_mat, R2_mat, prices, t, PI_markov_list, k, method_code):
    """Compute Markov belief update PI_markov for period t.

    Parameters
    ----------
    R1_mat, R2_mat : np.ndarray (NoP, NoP)
        Markov transition matrices.
    prices : np.ndarray
        Vector of possible prices.
    t : int
        Current period (1-based).
    PI_markov_list : list or np.ndarray
        If t==1: np.ndarray of shape (2,) — initial beliefs [pi_R1, pi_R2]
        If t>=2: list of [PI_R1, PI_R2] pairs for each past period
    k : np.ndarray (2, 2)
        Regime transition probabilities.
    method_code : int
        Solution method code.

    Returns
    -------
    list : [PI_markov1, PI_markov2]
        For t==1: [scalar, scalar]
        For t>=2: [np.ndarray, np.ndarray] of shape (NoP^(t-1), NoP)
    """
    NoP = len(prices)

    if t == 1:
        # For t=1, belief values are given directly
        return [PI_markov_list[0], PI_markov_list[1]]

    # Get past beliefs
    PI_past_R1 = PI_markov_list[t - 2][0]  # beliefs in t-1 for R1
    PI_past_R2 = PI_markov_list[t - 2][1]  # beliefs in t-1 for R2

    # For t==2, PI_past is scalar -> replicate to (1, NoP) matrix
    if t == 2:
        PI_past_R1 = np.full((1, NoP), PI_past_R1)
        PI_past_R2 = np.full((1, NoP), PI_past_R2)

    # Get initial beliefs for CEC fallback
    initial_pi_r1 = float(PI_markov_list[0][0])
    initial_pi_r2 = float(PI_markov_list[0][1])

    PI_m1, PI_m2 = _update_pi_markov_core(
        R1_mat, R2_mat, NoP, t, PI_past_R1, PI_past_R2,
        k, method_code, initial_pi_r1, initial_pi_r2
    )

    return [PI_m1, PI_m2]
